File: brainfuck2/bf.lua
local INC = 0
local MOVE = 1
local PRINT = 3
local LOOP = 4

local function Tape()
  local pos = 1;
  local data = {0};

  local get = function()
    return data[pos];
  end

  local inc = function(x)
    data[pos] = data[pos] + x;
  end

  local move = function(x)
    local length = #data;
    pos = pos + x;
    for i = length + 1, pos do
      data[i] = 0;
    end
  end

  return {
    get = get;
    inc = inc;
    move = move;
  };
end

local function Brainfuck(text)
  local function parse(source, i)
    local res = {};
    while i <= source:len() do
      local c = source:sub(i, i);
      if c == "+" then
        table.insert(res, {INC, 1});
      elseif c == "-" then
        table.insert(res, {INC, -1});
      elseif c == ">" then
        table.insert(res, {MOVE, 1});
      elseif c == "<" then
        table.insert(res, {MOVE, -1});
      elseif c == "." then
        table.insert(res, {PRINT});
      elseif c == "[" then
        local loop_code;
        loop_code, i = parse(source, i+1);
        table.insert(res, {LOOP, loop_code});
      elseif c == "]" then
        break;
      end
      i = i + 1;
    end
    return res, i;
  end

  local function _run(program, tape)
    for i = 1, #program do
      local op = program[i];
      local operator = op[1];
      if operator == INC then
        tape.inc(op[2]);
      elseif operator == MOVE then
        tape.move(op[2]);
      elseif operator == LOOP then
        while tape.get() > 0 do
          _run(op[2], tape);
        end
      elseif operator == PRINT then
        io.write(string.char(tape.get()));
        io.flush();
      end
    end
  end

  local function run()
    _run(parse(text, 1), Tape())
  end

  return {
    run = run;
  };
end

local socket = require "socket"
socket.protect(function()
  local c = socket.try(socket.connect("localhost", 9001))
  local try = socket.newtry(function() c:close() end)
  try(c:send(type(jit) == 'table' and "LuaJIT" or "Lua"))
  c:close()
end)()

local text = io.open(arg[1]):read("*a");
local brainfuck = Brainfuck(text);
brainfuck.run();

File: brainfuck2/bf.pl
#! /usr/bin/env perl

use 5.10.1;
use strict;
use warnings;
use Socket;
no warnings 'experimental';

package Op;

sub new {
    my $class = shift;
    my $op = shift;
    my $val = shift;

    my $self = {
        op => $op,
        val => $val,
    };
    bless $self, $class;
    return $self;
}


package Tape;

sub new {
    my $class = shift;

    my $self = {
        tape => [0],
        pos => 0,
    };
    bless $self, $class;
    return $self;
}

sub get {
    my $self = shift;

    return $self->{tape}[$self->{pos}];
}

sub inc {
    my $self = shift;
    my $x = shift;

    $self->{tape}[$self->{pos}] += $x;
}

sub move {
    my $self = shift;
    my $x = shift;

    $self->{pos} += $x;
    my $missing = $self->{pos} - @{$self->{tape}} + 1;
    for (my $i = 0; $i < $missing; $i++) {
        push @{$self->{tape}}, 0;
    }
}


package Main;

my $INC   = 1;
my $MOVE  = 2;
my $PRINT = 3;
my $LOOP  = 4;

sub parse {
    my $source = shift;
    my $i = shift || 0;

    my $repr = [];
    for (; $i < @{$source}; $i++) {
        given ($source->[$i]) {
            when ('+') { push @$repr, Op->new($INC,   1); }
            when ('-') { push @$repr, Op->new($INC,  -1); }
            when ('>') { push @$repr, Op->new($MOVE,  1); }
            when ('<') { push @$repr, Op->new($MOVE, -1); }
            when ('.') { push @$repr, Op->new($PRINT); }
            when ('[') {
                my ($parsed_loop, $new_i) = parse($source, $i + 1);
                $i = $new_i;
                push @$repr, Op->new($LOOP, $parsed_loop);
            }
            when (']') { last; }
        }
    }
    return ($repr, $i);
}

sub run {
    my $parsed = shift;
    my $tape = shift;
    foreach my $op (@$parsed) {
        CORE::given ($op->{op}) {
            when ($INC)   { $tape->inc($op->{val}); }
            when ($MOVE)  { $tape->move($op->{val}); }
            when ($PRINT) { printf "%c", $tape->get(); }
            when ($LOOP)  {
                while ($tape->get() > 0) {
                    run($op->{val}, $tape);
                }
            }
        }
    }
}

socket(my $socket, Socket::PF_INET, Socket::SOCK_STREAM, (getprotobyname('tcp'))[2]);
if (connect($socket, Socket::pack_sockaddr_in(9001, Socket::inet_aton('localhost')))) {
  print $socket "Perl";
}
close($socket);

open (FH, "<", shift) or die $!;
undef $/;
$| = 1;
my ($parsed, $n) = parse([split //, <FH>]);
my $tape = Tape->new();
run($parsed, $tape);
